### PR TITLE
MODOAIPMH-296: Update dependencies in mod-oai-pmh before Iris release

### PR DESCRIPTION
**Pull request by IlliaDaliek**
Jira: https://issues.folio.org/browse/MODOAIPMH-296

### PURPOSE
Update oai-pmh dependencies before Iris release. 

### CHANGES
**Dependencies updated:**
- vertx-completable-future removed
- mod-configuration-client 5.4.0 -> 5.6.0
- mod-source-record-storage-client.version 4.0.1 -> 4.1.4

The CompletableFuture replaced with vertx Future.

ModuleDescriptor will be updated at the release PR.

### CODE REVIEW

**Review by illia-borysenko — APPROVED** [review]

**Comment by IlliaDaliek** [comment]
The coverage is 73% but this coverage relates to the new code. OaiPmhSetImpl has changes that contain a correction of log messages and I don't think that here any further unit tests are needed.

**Review by Igor-Gorchakov — APPROVED** [review]

**Comment by craigmcnally** [comment]
Can you please explain why in some places were using log4j2 style lazy evaluation but in other places were using string.format %s for log messages?

**Comment by craigmcnally** [comment]
It would be helpful to have a high level description of what you're doing here with logging changes

**Comment by craigmcnally** [comment]
> Can you please explain why in some places were using log4j2 style lazy evaluation but in other places were using string.format %s for log messages?
> 
> Examples:
> https://github.com/folio-org/mod-oai-pmh/blob/ec594af1f2a9f186225f3ef8712eecd90bb60f22/src/main/java/org/folio/oaipmh/helpers/RepositoryConfigurationUtil.java#L67
> 
> And
> 
> https://github.com/folio-org/mod-oai-pmh/blob/ec594af1f2a9f186225f3ef8712eecd90bb60f22/src/main/java/org/folio/rest/impl/ModTenantAPI.java#L164
> 
> And here you're switching from one to the other, and as a slight variation this class has a static import of String.format
> https://github.com/folio-org/mod-oai-pmh/blob/ec594af1f2a9f186225f3ef8712eecd90bb60f22/src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java#L74
> 
> I'm _guessing_ that the lazy eval { } ones aren't doing what you think they are, and should instead use string.format until we upgrade to log4j2. In any event, we need to bring some consistency to this and ensure log messages are actually being properly formed. It's not very helpful to have a bunch of curly braces all over the place in the log ;)

Looking around s bit I see we are using log4j2, or at least we have log4j2.properties files.  If that's the case why wouldn't we use lazy evaluation instead of String.format?

This is confusing to me. I'll be quiet and give you a chance to explain what's going on here.

**Comment by IlliaDaliek** [comment]
> > Can you please explain why in some places were using log4j2 style param substitution/lazy evaluation but in other places were using string.format %s for log messages?
> > Examples:
> > https://github.com/folio-org/mod-oai-pmh/blob/ec594af1f2a9f186225f3ef8712eecd90bb60f22/src/main/java/org/folio/oaipmh/helpers/RepositoryConfigurationUtil.java#L67
> > 
> > And
> > https://github.com/folio-org/mod-oai-pmh/blob/ec594af1f2a9f186225f3ef8712eecd90bb60f22/src/main/java/org/folio/rest/impl/ModTenantAPI.java#L164
> > 
> > And here you're switching from one to the other, and as a slight variation this class has a static import of String.format
> > https://github.com/folio-org/mod-oai-pmh/blob/ec594af1f2a9f186225f3ef8712eecd90bb60f22/src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java#L74
> > 
> > I'm _guessing_ that the lazy eval { } ones aren't doing what you think they are, and should instead use string.format until we upgrade to log4j2. In any event, we need to bring some consistency to this and ensure log messages are actually being properly formed. It's not very helpful to have a bunch of curly braces all over the place in the log ;)
> 
> Looking around s bit I see we are using log4j2, or at least we have log4j2.properties files. If that's the case why wouldn't we use the log4j2 param substitution/lazy evaluation instead of String.format?
> 
> This is confusing to me. I'll be quiet and give you a chance to explain what's going on here.


Hi @craigmcnally. I replaced the circle braces with "String.format(..)" because they weren't replaced with values passed to log method parameters. But I figured out that we should put the value starting from zero to circle braces to point a position for values to be put in the log message. I did such refactoring among the code solving this issue. As well in some places was an incorrect order of parameters for error logs, firstly we should pass a throwable object and then objects to be put in the log message. I tested these changes locally to ensure that the logs are correct.

**Comment by craigmcnally** [comment]
> Hi @craigmcnally. I replaced the circle braces with "String.format(..)" because they weren't replaced with values passed to log method parameters. But I figured out that we should put the value starting from zero to circle braces to point a position for values to be put in the log message. I did such refactoring among the code solving this issue. As well in some places was an incorrect order of parameters for error logs, firstly we should pass a throwable object and then objects to be put in the log message. I tested these changes locally to ensure that the logs are correct.

OK, I think I see what's happening here... the `{N}` syntax was a hint.  By default Vertx uses JUL logging.  If you don't specify an override via `vertx.logger-delegate-factory-class-name` that's what will be used.  If you look at the pom file you can see that we specify `io.vertx.core.logging.Log4j2LogDelegateFactory`:  https://github.com/folio-org/mod-oai-pmh/blob/90ad79589b58c83c8c5e722c93e07a1123a560c1/pom.xml#L664

Is it the case where you're running your unit tests via an IDE which doesn't provide this system property, and therefore you get the JUL logging?

**Comment by IlliaDaliek** [comment]
> > Hi @craigmcnally. I replaced the circle braces with "String.format(..)" because they weren't replaced with values passed to log method parameters. But I figured out that we should put the value starting from zero to circle braces to point a position for values to be put in the log message. I did such refactoring among the code solving this issue. As well in some places was an incorrect order of parameters for error logs, firstly we should pass a throwable object and then objects to be put in the log message. I tested these changes locally to ensure that the logs are correct.
> 
> OK, I think I see what's happening here... the `{N}` syntax was a hint. By default Vertx uses JUL logging. If you don't specify an override via `vertx.logger-delegate-factory-class-name` that's what will be used. If you look at the pom file you can see that we specify `io.vertx.core.logging.Log4j2LogDelegateFactory`:
> 
> https://github.com/folio-org/mod-oai-pmh/blob/90ad79589b58c83c8c5e722c93e07a1123a560c1/pom.xml#L664
> 
> Is it the case where you're running your unit tests via an IDE which doesn't provide this system property, and therefore you get the JUL logging?


@craigmcnally , I verified the logs by real calls to mod-oai-pmh. I used run IDE configuration and I didn't specify thevertx logger property.
image

**Comment by craigmcnally** [comment]
> I should provide a Log4j2LogDelegateFactory in option and adjust the log messages to use it. Am I correct?

That's right

**Comment by IlliaDaliek** [comment]
> > I should provide a Log4j2LogDelegateFactory in option and adjust the log messages to use it. Am I correct?
> 
> That's right


Hi @craigmcnally, I set up a run configuration to use Log4j2LogDelegateFactory and corrected the log messages accordingly. From those things that you already viewed I refactored a couple of lines to use circle braces instead of String.format in logs at ModTenantApi class, and added a method to transform the Java dto to json string in order to log a real JSON body instead of the object name in java memory (like org.folio.rest.jaxrs.model.FolioSet@2234).

File: src/main/java/org/folio/oaipmh/helpers/AbstractGetRecordsHelper.java
Comment (craigmcnally) [inline src/main/java/org/folio/oaipmh/helpers/AbstractGetRecordsHelper.java]: Where this is an error it's unlikely to make a big difference, but generally its better to use lazy logging...

```
logger.error("{} response from SRS status code: {}: {}",
    () -> request.getVerb().value(),
    () -> response.statusMessage(),
    () -> response.statusCode());
```

File: src/main/java/org/folio/rest/impl/OaiPmhImpl.java
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/OaiPmhImpl.java]:
```suggestion
                logger.debug("{} response: {}", verb,  () -> response.getEntity());
```

File: src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java:76]:
```suggestion
        logger.info("Put set by id with id: '{}' and body: {}", id, () -> entityToJsonString(entity));
```

File: src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java:95]:
```suggestion
        logger.info("Post set with body: {}", () -> entityToJsonString(entity));
```

**Review by craigmcnally — APPROVED** [review]
A few suggestions,.  Approving now so we don't need another review iteration

File: src/main/java/org/folio/oaipmh/helpers/AbstractGetRecordsHelper.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/oaipmh/helpers/AbstractGetRecordsHelper.java]: Okay, thank you for the advice!

File: src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java
Comment (IlliaDaliek) [reply src/main/java/org/folio/rest/impl/OaiPmhSetImpl.java:95]: I tried the approach but the compiler swears on it.

I will back to it at the bugfix ticket PR's

**Comment by craigmcnally** [comment]
I see you chose not to commit my suggestions, and I wanted to make sure you understand the implication of these changes as they may impact performance.  https://github.com/folio-org/mod-oai-pmh/pull/196#discussion_r595140129 for instance, that encoding JSON as a string will happen all the time, even if debug logging isn't enabled.  By calling the method via a lambda, it would only be executed if debug logging was enabled.  Where these responses are potentially massive, this could be a non-trivial operation.  Furthermore, it doesn't make sense to mix plain old string concatenation with parameter substitution.  

I had approved to avoid needing another review cycle, but was expecting the changes to be committed prior to you merging the PR.  I can see now that expectation wasn't clear.

**Pull request merged**
